### ISSUE TITLE
Hashable and tuple conflict with each other

### ISSUE THREAD

**Issue by madig** [microsoft/pyright#3814]
<h3> Environment data </h3>

-   Language Server version: 2022.8.10
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 3.8.10
-   python.analysis.indexing: null
-   python.analysis.typeCheckingMode: basic

<h3> Code Snippet </h3>

```python
# pyright: strict

from __future__ import annotations

from typing import Callable, Hashable


def foo(
    number: int,
    function: Callable[[int], Hashable | list[Hashable] | set[Hashable] | tuple[Hashable] | None],
) -> dict[Hashable, set[str]]:
    sets: dict[Hashable, set[str]] = {}

    things = function(number)
    if things is None:
        pass
    elif isinstance(things, (list, set, tuple)):
        # reveal_type(things)
        for element in things:
            sets.setdefault(element, set()).add("x")
    else:
        sets.setdefault(things, set()).add("x")

    return sets

```

<h3> Repro Steps </h3>

1. Run Pylance over the above file.

<h3> Expected behavior </h3>

`element` should be detected as any of the three object types in the `isinstance` check, i.e. `things: list[Hashable] | set[Hashable] | tuple[Hashable]` and `element: Hashable`.

<h3> Actual behavior </h3>

Both Pylance and mypy detect `element` to be `Any | Hashable`, but Pylance complains about an unknown argument type.

I understand that a tuple is `Hashable`, too, but shouldn't the `isinstance` check be enough to disambiguate the type?

Note: Weirdly enough, if I try the snippet in a fresh VSCode window, the error is different; Pylance just complains about an unknown argument type. In my other workspace, the commented out `reveal_type(things)` says:

* Pylance: Type of "things" is "tuple[Unknown, ...] | list[Hashable] | set[Hashable] | tuple[Hashable]"
* mypy(note): Revealed type is "Union[builtins.tuple[Any], builtins.list[typing.Hashable], builtins.set[typing.Hashable]]"

<h3> Logs </h3>

```
[Info  - 19:12:37] (15308) Pylance language server 2022.8.10 (pyright 6fa755c0) starting
[Info  - 19:12:37] (15308) Server root directory: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist
[Info  - 19:12:37] (15308) Starting service instance "tttt"
[Info  - 19:12:37] (15308) Notebook support: Legacy
[Info  - 19:12:37] (15308) Interactive window support: Legacy
[Info  - 19:12:37] (15308) No configuration file found.
[Info  - 19:12:37] (15308) No pyproject.toml file found.
[Info  - 19:12:37] (15308) Setting pythonPath for service "tttt": "C:\Program Files\Python38\python.exe"
[Warn  - 19:12:37] (15308) stubPath c:\Users\user\AppData\Local\Temp\tttt\typings is not a valid directory.
[Info  - 19:12:37] (15308) Assuming Python version 3.8
[Info  - 19:12:37] (15308) Assuming Python platform Windows
[Info  - 19:12:37] (15308) Search paths for c:\Users\user\AppData\Local\Temp\tttt
[Info  - 19:12:37] (15308)   c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib
[Info  - 19:12:37] (15308)   c:\Users\user\AppData\Local\Temp\tttt
[Info  - 19:12:37] (15308)   c:\Users\user\AppData\Local\Temp\tttt\typings
[Info  - 19:12:37] (15308)   c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stubs\...
[Info  - 19:12:37] (15308)   c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\bundled\stubs
[Info  - 19:12:37] (15308)   c:\Program Files\Python38\DLLs
[Info  - 19:12:37] (15308)   c:\Program Files\Python38\Lib
[Info  - 19:12:37] (15308)   c:\Program Files\Python38
[Info  - 19:12:37] (15308)   c:\Users\user\AppData\Roaming\Python\Python38\site-packages
[Info  - 19:12:37] (15308)   c:\Program Files\Python38\Lib\site-packages
[Info  - 19:12:37] (15308) Adding fs watcher for library directories:
 C:\Program Files\Python38\DLLs
C:\Program Files\Python38\lib
C:\Program Files\Python38
C:\Users\user\AppData\Roaming\Python\Python38\site-packages
C:\Program Files\Python38\lib\site-packages
[Info  - 19:12:37] (15308) Adding fs watcher for directories:
 c:\Users\user\AppData\Local\Temp\tttt
[Info  - 19:12:37] (15308) Searching for source files
[Info  - 19:12:37] (15308) Found 1 source file
(15308) [IDX(FG)] index libraries c:\Users\user\AppData\Local\Temp\tttt (index) ...
(15308) [IDX(FG)]   read stdlib indices (16ms)
(15308) [IDX(FG)] index libraries c:\Users\user\AppData\Local\Temp\tttt (index) [succeed] (16ms)
(15308) [FG] parsing: c:\Users\user\AppData\Local\Temp\tttt\a.py (18ms)
(15308) [FG] parsing: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 2ms] (52ms)
(15308) [FG] binding: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\builtins.pyi (20ms)
(15308) [FG] binding: c:\Users\user\AppData\Local\Temp\tttt\a.py (1ms)
[Info  - 19:12:37] (15308) Background analysis(1) root directory: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist
[Info  - 19:12:37] (15308) Background analysis(1) started
(15308) Background analysis message: setConfigOptions
(15308) Background analysis message: setImportResolver
(15308) Background analysis message: ensurePartialStubPackages
(15308) Background analysis message: setTrackedFiles
(15308) Background analysis message: markAllFilesDirty
(15308) Background analysis message: setFileOpened
(15308) Background analysis message: analyze
(15308) [BG(1)] analyzing: c:\Users\user\AppData\Local\Temp\tttt\a.py ...
(15308) [BG(1)]   parsing: c:\Users\user\AppData\Local\Temp\tttt\a.py (16ms)
(15308) [BG(1)]   parsing: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 2ms] (51ms)
(15308) [BG(1)]   binding: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\builtins.pyi (23ms)
(15308) [BG(1)]   binding: c:\Users\user\AppData\Local\Temp\tttt\a.py (1ms)
(15308) [BG(1)]   checking: c:\Users\user\AppData\Local\Temp\tttt\a.py ...
(15308) [BG(1)]     parsing: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\typing_extensions.pyi [fs read 1ms] (4ms)
(15308) [BG(1)]     binding: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\typing_extensions.pyi (1ms)
(15308) [BG(1)]     parsing: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\typing.pyi [fs read 1ms] (9ms)
(15308) [BG(1)]     binding: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\typing.pyi (5ms)
(15308) [BG(1)]     parsing: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 0ms] (8ms)
(15308) [BG(1)]     binding: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (5ms)
(15308) [BG(1)]     parsing: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\abc.pyi [fs read 1ms] (2ms)
(15308) [BG(1)]     binding: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\abc.pyi (0ms)
(15308) [BG(1)]     parsing: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\__future__.pyi [fs read 0ms] (0ms)
(15308) [BG(1)]     binding: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\__future__.pyi (1ms)
(15308) [BG(1)]     parsing: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\collections\abc.pyi [fs read 0ms] (0ms)
(15308) [BG(1)]     binding: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\collections\abc.pyi ...
(15308) [BG(1)]       parsing: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\_collections_abc.pyi [fs read 0ms] (1ms)
(15308) [BG(1)]       binding: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\_collections_abc.pyi (1ms)
(15308) [BG(1)]     binding: c:\Users\user\.vscode\extensions\ms-python.vscode-pylance-2022.8.10\dist\typeshed-fallback\stdlib\collections\abc.pyi (2ms)
(15308) [BG(1)]   checking: c:\Users\user\AppData\Local\Temp\tttt\a.py (89ms)
(15308) [BG(1)] analyzing: c:\Users\user\AppData\Local\Temp\tttt\a.py (181ms)
(15308) Background analysis message: resumeAnalysis

```

**Comment by erictraut** [comment]
After looking at this more closely, I think pyright is doing the right thing.

The issue is that the `tuple` class implements the `Hashable` protocol. (By contrast, `list` and `set` do not.) 

Consider the following simplified example:

```python
def func(x: Hashable | tuple[Hashable]):
    if isinstance(x, tuple):
        reveal_type(x) # tuple[Unknown, ...] | tuple[Hashable]

func((1, 2, 3))
```

This type checks fine because the expression `(1, 2, 3)` is type `tuple[int, int, int]`, which is type compatible with `Hashable`. The `isinstance` check in `func` correctly narrows the type of `x` to `tuple[Unknown, ...] | tuple[Hashable]`.

So I don't consider this a bug in pyright. If you want to use strict type checking, you'll need to find a workaround.

**Comment by madig** [comment]
Ok, thanks for checking. A shame, but I guess it's the nature of a broad "hashable" type.
